### PR TITLE
Add the combination engine

### PR DESCRIPTION
**Pull request by ajreynol**
This adds the combination engine, which is the module of TheoryEngine which implements the `combineTheories` method and owns the various components of theory combination, which includes equality engine manager, model manager, and the "shared solver" (to come later).  It will have two variants, `CombinationCareGraph` and `CombinationModelBased`, the former is added with this PR.

FYI @barrettcw 

The next PR will connect this module to TheoryEngine and remove a few existing methods from TheoryEngine, as they are implemented in the modules of this class.

### CODE REVIEW

File: src/theory/combination_care_graph.h [hunk 1]
```diff
@@ -0,0 +1,53 @@
...
+ **
+ ** \brief Management of a care graph based approach for theory combination.
+ **/
+
+#include "cvc4_private.h"
+
+#ifndef CVC4__THEORY__COMBINATION_CARE_GRAPH__H
+#define CVC4__THEORY__COMBINATION_CARE_GRAPH__H
+
+#include <map>
```
Comment (4tXJ7f) [inline src/theory/combination_care_graph.h:20]:
```suggestion
#include <vector>
```

File: src/theory/combination_care_graph.h [hunk 2]
```diff
@@ -0,0 +1,53 @@
...
+#include <memory>
```
Comment (4tXJ7f) [inline src/theory/combination_care_graph.h:21]:
```suggestion
```

File: src/theory/combination_care_graph.h [hunk 3]
```diff
@@ -0,0 +1,53 @@
...
+
+namespace CVC4 {
+
+class TheoryEngine;
+
+namespace theory {
+
+/**
+ * Manager for doing theory combination using care graphs. This is typically
+ * done via a distributed equalty engine architecture.
```
Comment (4tXJ7f) [inline src/theory/combination_care_graph.h:33]:
```suggestion
 * done via a distributed equality engine architecture.
```

File: src/theory/combination_care_graph.h [hunk 4]
```diff
@@ -0,0 +1,53 @@
...
+ */
+class CombinationCareGraph : public CombinationEngine
+{
+ public:
+  CombinationCareGraph(TheoryEngine& te,
+                       const std::vector<Theory*>& paraTheories);
+  ~CombinationCareGraph();
+
+  /** Build model */
```
Comment (4tXJ7f) [inline src/theory/combination_care_graph.h:42]:
```suggestion
```

File: src/theory/combination_engine.h [hunk 5]
```diff
@@ -0,0 +1,122 @@
...
+ **
+ ** \brief Abstract interface for theory combination.
+ **/
+
+#include "cvc4_private.h"
+
+#ifndef CVC4__THEORY__COMBINATION_ENGINE__H
+#define CVC4__THEORY__COMBINATION_ENGINE__H
+
+#include <map>
```
Comment (4tXJ7f) [inline src/theory/combination_engine.h:20]:
```suggestion
#include <vector>
```

File: src/theory/combination_care_graph.cpp [hunk 6]
```diff
@@ -0,0 +1,93 @@
...
+  }
+
+  Trace("combineTheories")
+      << "TheoryEngine::combineTheories(): care graph size = "
+      << careGraph.size() << std::endl;
+
+  // Now add splitters for the ones we are interested in
+  CareGraph::const_iterator care_it = careGraph.begin();
+  CareGraph::const_iterator care_it_end = careGraph.end();
+
```
Comment (4tXJ7f) [inline src/theory/combination_care_graph.cpp:52]:
```suggestion
```

File: src/theory/combination_care_graph.cpp [hunk 7]
```diff
@@ -0,0 +1,93 @@
...
+  prop::PropEngine* propEngine = d_te.getPropEngine();
+  for (; care_it != care_it_end; ++care_it)
+  {
+    const CarePair& carePair = *care_it;
+
```
Comment (4tXJ7f) [inline src/theory/combination_care_graph.cpp:57]:
```suggestion
  for (const CarePair& carePair : careGraph)
  {
```

File: src/theory/combination_care_graph.cpp (hunk 6 above)
Comment (ajreynol) [reply src/theory/combination_care_graph.cpp:52]: This is necessary in the for loop below.

**Pull request merged**